### PR TITLE
JTJ-56 Authentication

### PR DESCRIPTION
**Pull request by tgi-wtag**
This is my PR regarding Admin user Rights where a Admin can give a User certain rights.

The demo video is short and there isn't much to see, but it shows the functionality. 

https://github.com/tgi-wtag/Journey-tg/assets/141820846/a1a176e8-c5e3-41a1-8696-0e44405dc395

Im happy if you have any suggestions or Improvements, thank you in advance for reviewing my PR. 🌻

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -12,9 +18,26 @@ def create
     end
   end
 
+  def update_role
+    @user = User.find(params[:id])
+    
+    if @user.user?
+      @user.update(role: 'admin')
+      puts @user.errors.full_messages
```
Comment (chm-wtag) [inline app/controllers/users_controller.rb:26]: We never print logs like this. You can study about [rails logger](https://guides.rubyonrails.org/debugging_rails_applications.html#the-logger) and use that if you wish.

File: spec/controllers/users_controller_spec.rb [hunk 2]
```diff
@@ -17,10 +17,47 @@
...
+      post :create, params: { user: valid_params }
+      expect(response).to redirect_to(root_path)
+    end
+  end
+
+  describe 'PATCH #update_role' do
+    let(:user) { create(:user, role: 'admin') }
+
+    it 'when updating the role to user updates the user role to user' do
+      patch :update_role, params: { id: user.id, role: 'user' }
```
Comment (chm-wtag) [inline spec/controllers/users_controller_spec.rb:40]: let's use symbol

File: spec/controllers/users_controller_spec.rb [hunk 3]
```diff
@@ -17,10 +17,47 @@
...
+    it 'when updating the role to user sets a flash notice when updating to user' do
+      patch :update_role, params: { id: user.id, role: 'user' }
+      expect(flash[:notice]).to eq(I18n.t('user_role_updated_user'))
+    end
+
+    it 'when the role update fails does not change the user role' do
+      allow_any_instance_of(User).to receive(:update).and_return(false)
+      patch :update_role, params: { id: user.id, role: 'user' }
+      user.reload
+      expect(user.role).to eq('admin')
```
Comment (chm-wtag) [inline spec/controllers/users_controller_spec.rb:54]:
```suggestion
      expect(user.reload.role).to eq('admin')
```

**Review by chm-wtag — APPROVED** [review]

File: app/models/ability.rb [hunk 4]
```diff
@@ -2,29 +2,13 @@ class Ability
...
-    #
-    # The third argument is an optional hash of conditions to further filter the
-    # objects.
-    # For example, here the user can only update published articles.
-    #
-    #   can :update, Article, published: true
-    #
-    # See the wiki for details:
-    # https://github.com/CanCanCommunity/cancancan/blob/develop/docs/define_check_abilities.md
+    if user.admin?
```
Comment (evk-wtag) [inline app/models/ability.rb:5]: Let's add the corresponding rspec

File: db/migrate/20231106131223_change_role_type_in_users.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class ChangeRoleTypeInUsers < ActiveRecord::Migration[7.0]
+  def change
+    change_column :users, :role, 'integer USING CAST(role AS integer)'
```
Comment (evk-wtag) [inline db/migrate/20231106131223_change_role_type_in_users.rb:3]: Can you try 
```suggestion
    change_column :users, :role, :integer, using: 'role::integer'
```

File: db/schema.rb [hunk 6]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.0].define(version: 2023_10_13_083210) do
+ActiveRecord::Schema[7.0].define(version: 2023_11_06_131223) do
```
Comment (evk-wtag) [inline db/schema.rb:13]: No changes in schema after the migration? 🤔

File: db/migrate/20231106131223_change_role_type_in_users.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class ChangeRoleTypeInUsers < ActiveRecord::Migration[7.0]
```
Comment (evk-wtag) [inline db/migrate/20231106131223_change_role_type_in_users.rb:1]: Role was already integer in db right?

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -12,9 +18,26 @@ def create
     end
   end
 
+  def update_role
+    @user = User.find(params[:id])
+    
+    if @user.user?
+      @user.update(role: 'admin')
```
[the remark this answers is not in the document]
Comment (tgi-wtag) [reply app/controllers/users_controller.rb:25]: Yes it is, i tried it, but it didnt worked i dont know why 🤷🏻‍♂️

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (tgi-wtag) [reply app/controllers/users_controller.rb:26]: ok thank you very much for providing this for me, i will take a look at it and probably delete this spec

File: db/migrate/20231106131223_change_role_type_in_users.rb (hunk 7 above)
Comment (tgi-wtag) [reply db/migrate/20231106131223_change_role_type_in_users.rb:1]: yes i thought maybe it isnt, but it was, i will delete the migration

File: db/schema.rb (hunk 6 above)
Comment (tgi-wtag) [reply db/schema.rb:13]: yes wrong migration

File: app/models/ability.rb (hunk 4 above)
Comment (tgi-wtag) [reply app/models/ability.rb:5]: i added them

File: app/controllers/users_controller.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (tgi-wtag) [reply app/controllers/users_controller.rb:25]: i changed it it works now

File: spec/controllers/users_controller_spec.rb (hunk 2 above)
Comment (tgi-wtag) [reply spec/controllers/users_controller_spec.rb:40]: ok thank you

File: db/schema.rb [hunk 9]
```diff
@@ -10,7 +10,7 @@
...
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
```
Comment (lra-wtag) [inline db/schema.rb:16]: this should not change as there are no migrations

**Review by lra-wtag — APPROVED** [review]

File: db/schema.rb (hunk 9 above)
Comment (tgi-wtag) [reply db/schema.rb:16]: yes i know, there was a migration, which i used wrongfully and then deleted it, evea already pointed it out, thats why i deleted the migration, but thanks for informing

**Pull request merged**